documentation/src/test/java/example/registration/WebServerDemo.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+package example.registration
+
+import org.junit.jupiter.api.Assertions.assertEquals
+
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.BeforeAllCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+import org.junit.jupiter.api.extension.RegisterExtension
+
+// tag::user_guide[]
+class WebServerDemo {
+    // end::user_guide[]
+    // @formatter:off
+    // tag::user_guide[]
+    companion object {
+        @JvmField
+        @RegisterExtension
+        val server = WebServerExtension.builder()
+                .enableSecurity(false)
+                .build()
+    }
+    // end::user_guide[]
+    // @formatter:on
+    // tag::user_guide[]
+
+    @Test
+    fun getProductList() {
+        // end::user_guide[]
+        // tag::user_guide[]
+        // Use WebClient to connect to web server using serverUrl and verify response
+        val webClient = WebClient()
+        val serverUrl = server.serverUrl
+        assertEquals(200, webClient.get("$serverUrl/producs").responseStatus)
+    }
+}
+// end::user_guide[]
+
+class WebServerExtension : BeforeAllCallback {
```

Comment:
Can we reuse the Java equivalents of these classes?

Reply:
👍

Reply:
Hi,

regarding this and other comments. I merely ported the java extension to kotlin. I thought I could reuse the Java classes, but am not sure how I can achieve that without name clashes (i.e. the `WebServerDemo` class has the same name in both the Java and Kotlin examples).

I am not familiar with how JUnit 5 itself is developed, I noticed that the examples are themselves tested, but don't know whether this Kotlin example is tested too.

Happy to make the changes needed with some guidance.

Reply:
> I am not familiar with how JUnit 5 itself is developed, I noticed that the examples are themselves tested, but don't know whether this Kotlin example is tested too.

Yes, the Kotlin demos are also tested in the build. Note the inclusion of `*.Demo` classes below.

https://github.com/junit-team/junit5/blob/35732161b6fc20d5825c1c049345f0e0e2eacd9e/documentation/documentation.gradle#L17-L30

Reply:
E.g. if I merely remove the `WebServerExtension`, `WebClient` and `WebResponse` classes the `:documentation:compileTestKotlin` task fails with "Unresolved reference" errors.

Reply:
To avoid the naming conflict, please rename your Kotlin demo class to `WebServerKotlinDemo`.

Reply:
Otherwise, you _should_ be able to reuse the `WebServerExtension`, `WebClient`, and `WebResponse` Java types. They are all top-level classes in the `example.registration` package.

Reply:
> Happy to make the changes needed with some guidance.

Is that sufficient guidance? 😉

Reply:
> if I merely remove the `WebServerExtension`, `WebClient` and `WebResponse` classes the `:documentation:compileTestKotlin` task fails with "Unresolved reference" errors.

I'm not a Kotlin expert, but perhaps you need to move those classes to their own files and make them `public` types.

If that doens't help, perhaps there is an issue regarding the order in which Java and Kotlin test classes are compiled.

Reply:
Actually, now that I think about it, I think we do in fact compile Kotlin classes before Java classes.

So that would explain why the Java types are not visible during the Kotlin test class compilation task execution.

Maybe we need to remove those types from the Java demo, keep them in the Kotlin demo, and then reference the Kotlin types from the Java demo?

If that proves to be too much work, we can also just live with the duplication. After all, they are only dummy examples. 😉

Reply:
I came to the same conclusion!

I tried to add `compileTestJava` as a dependency to `compileTestKotlin` and gradle failed saying that I had introduced a circular dependency, so I realized that I should keep the classes in Kotlin and remove the Java ones.

I was a bit surprised about this dependency though, I would have expected the inverse.

The file name should not matter as that is a "compilation unit" in Kotlin, but the class should definitely be renamed.

I am going to give a shot at removing the classes from the java file and keeping them in the kotlin one.